### DISCUSSION TITLE
How to create multiple applications in single test run

### DISCUSSION THREAD

**Discussion by IngvarX** [AvaloniaUI/Avalonia#5688] [Q&A] [answered]
Hello!

Is there a way to create multiple instances of application in single test run? I know that desktop lifetime requires single instance of app so let's say I will execute tests one by one, But this code fails if execute it twice:

```charp
AppBuilder
                .Configure<App>()
                .UsePlatformDetect()
                .UseReactiveUI();
    }
```

Any ideas how to solve this?

**Comment by maxkatz6** [comment, marked as answer]
I don't think it possible to have multiple app instances in one process, since avalonia uses static services locator and dispatcher.

Also it might not be possible on some platforms to run multiple UI threads at all (see: macos), but this specific problem can be solved with Headless platform, starting tests without any UI and rendering, but with all visual elements and ability to raise input events.

**Reply by maxkatz6** [reply]
It should be possible to isolate apps with app domains on .net framework or use different processes for testing with some layer of communication, which will be quite complex in the end.

**Reply by maxkatz6** [reply]
May I ask why you need to test multiple Avalonia instances inside of one test process?

**Reply by maxkatz6** [reply]
On, nevermind. I just saw that you don't actually need to run multiple avalonia instances in parallel for single test.
So you should be able to use approach like we have in our repository for controls testing, that Gillibald proposed - https://github.com/AvaloniaUI/Avalonia/blob/master/tests/Avalonia.UnitTests/UnitTestApplication.cs

Or use newer Headless platform, example https://github.com/kekekeks/Avalonia-unit-testing-with-headless-platform

**Reply by IngvarX** [reply]
I use headless platform but it still doesn't work because I can't create new app/ reuse previous one

**Reply by maxkatz6** [reply]
Can you share code, how did you set up headless platform?

For xUnit you can simply reuse this class, just update initialization code and attribute namespaces 
https://github.com/kekekeks/Avalonia-unit-testing-with-headless-platform/blob/master/HeadlessTests/AvaloniaTestFramework.cs

**Reply by IngvarX** [reply]
https://github.com/IngvarX/Camelot/blob/F7-test/tests/Camelot.Ui.Tests/AvaloniaApp.cs

**Reply by maxkatz6** [reply]
Have you tried to use approach from repository I used as an example? With integration with xUnit.
Or does it has limitations for you?

**Reply by IngvarX** [reply]
Not yet. Anyway it looks like it's used for unit testing while I'm trying to do integration testing with fully functional app

**Reply by IngvarX** [reply]
Checked code, but seems that it doesn't create lifetime that I need to get main window

**Reply by IngvarX** [reply]
Ok, after some struggling I was able to separate tests. Thanks!

**Comment by Gillibald** [comment]
https://github.com/AvaloniaUI/Avalonia/blob/master/tests/Avalonia.UnitTests/UnitTestApplication.cs
All unit tests use that technique

**Reply by IngvarX** [reply]
what about headless platform multiple UI tests with each own app initialization?
